carbon/job.py

```diff
@@ -75,40 +84,113 @@ class Job:
+            else:
+                raise ValueError(f"Failed to fetch job data: {e}")
+
+        sub_jobs = []
+        for row in output.stdout.splitlines()[3:]:
+            items = row.split()
+            label = items[0]
+            state = items[4]
+            # Get all the subjobs which are running, finished, or expired (finished but
+            # other subjobs are still running).
+            if re.fullmatch(r"^\d+\[\d+\](?:..*)?$", label) and state in [
+                "R",
+                "F",
+                "X",
+            ]:
+                # Add subjobs without server label to improve consistency
+                sub_jobs.append(label.split(".")[0])
+
+        return sub_jobs
+
+    @classmethod
+    def from_PBS_bulk(cls, ids: list[str], ignore_failed: bool = False) -> list[Self]:
+        """Create a list of Job objects by fetching data from PBS for multiple job IDs.
 
         Args:
-            id (str): The job identifier to fetch from the scheduler.
+            ids (list[str]): The job identifiers to fetch from the scheduler.
+            ignore_failed (bool): If True, don't crash out when jobs cannot be parsed
+                but don't add to the job list.
 
         Returns:
-            Job: An instance of the Job class populated with scheduler data.
+            list[Job]: A list containing instances of the Job class corresponding to
+                each of the ids.
 
         Raises:
             ValueError: If fetching or parsing job data fails, or if no job data is
                 found.
             UnknownJobIDError: If PBS returns exit code 153 for unknown job ID.
-            MalformedJobIDError: If the job ID is not formatted correctly.
-            JobStateError: If the job is in an invalid state.
+            MalformedJobIDError: If some of the job IDs are not formatted correctly.
+            JobStateError: If some of the jobs are in an invalid state.
             NotImplementedError: If the memory format is not supported.
         """
-        # Job ID should either be digits only or digits plus an index in square brackets
-        if not re.fullmatch(r"\d+(\[\d+\])?", id):
+        malformed_ids = []
+        for id in ids:
+            if not re.fullmatch(r"^\d+(\[\d+\])?(?:..*)?$", id):
+                malformed_ids.append(id)
+        if malformed_ids and not ignore_failed:
             raise MalformedJobIDError(
-                f"Malformed job ID: {id}. Should contain only digits"
+                "Malformed job ID(s): " + " ".join(malformed_ids) + ". "
+                "Should be composed of digits, "
+                "optionally followed by an index in square brackets, "
+                "optionally followed by a full stop and the PBS server name."
             )
+        elif malformed_ids and ignore_failed:
+            # Remove malformed ids from the list of ids
+            ids = [id for id in ids if id not in malformed_ids]
```

Comment:
Overall I wonder how much it's worth doing these checks as they have to the potential to be fairly brittle potentially even between systems using the same scheduler. The ultimate arbiter of whether an id is well formed is the scheduler so why do additional checks?